### PR TITLE
Install a global init script to collect HMS lineage for a few weeks.

### LINKED ISSUE #324: Install a global init script to collect HMS lineage for a few weeks

### PR DESCRIPTION
**Pull request by dipankarkush-db**
Fixes #324

### CODE REVIEW

**Comment by dipankarkush-db** [comment]
[
image
](url)


image

File: src/databricks/labs/ucx/hive_metastore/hms_lineage.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+cat << EOF >>  $driver_conf
+  [driver] {
+   "spark.databricks.dataLineage.enabled" = true
+   }
+EOF
+fi"""
+        return base64.b64encode(_init_script_content.encode()).decode()
+
+    def _add_global_init_script(self):
+        self._ws.global_init_scripts.create(
```
Comment (FastLee) [inline src/databricks/labs/ucx/hive_metastore/hms_lineage.py:26]: Do we check for existing init scripts? Can we append to existing scripts? Is there a risk?

**Review by FastLee — CHANGES_REQUESTED** [review]

**Review by william-conti — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/runtime.py [hunk 2]
```diff
@@ -187,6 +188,14 @@ def crawl_mounts(cfg: WorkspaceConfig):
     mounts.inventorize_mounts()
 
 
+@task("assessment", depends_on=[setup_schema])
+def enable_hms_lineage(cfg: WorkspaceConfig):
```
Comment (nfx) [inline src/databricks/labs/ucx/runtime.py:192]: It'll be a better fit if the init script is rolled out in the installer. It'll be simpler to support. Ask it in the installation flow as a question

File: src/databricks/labs/ucx/hive_metastore/hms_lineage.py (hunk 1 above)
Comment (nfx) [reply src/databricks/labs/ucx/hive_metastore/hms_lineage.py:26]: +1, it's an easy add

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/hive_metastore/hms_lineage.py (hunk 1 above)
Comment (dipankarkush-db) [reply src/databricks/labs/ucx/hive_metastore/hms_lineage.py:26]: Hi @FastLee and @nfx - are you referring to adding the portion in an existing **Global** init scripts?

File: src/databricks/labs/ucx/hive_metastore/hms_lineage.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        return base64.b64encode(_init_script_content.encode()).decode()
+
+    def _add_global_init_script(self) -> CreateResponse:
+        created_script = self._ws.global_init_scripts.create(
+            name=f"hms-lineage-{time.time_ns()}", script=self._get_init_script_content(), enabled=True
+        )
+        return created_script
+
+    def add_spark_config_for_hms_lineage(self):
+        created_script = self._add_global_init_script()
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/hms_lineage.py:33]: Add a mandatory check for any global init script with "spark.databricks.dataLineage.enabled" string to skip the addition of this init script

File: src/databricks/labs/ucx/hive_metastore/hms_lineage.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        self._add_sql_wh_config()
+        return created_script.script_id
+
+    def _add_sql_wh_config(self):
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/hms_lineage.py:37]: Don't add empty methods, they confuse people while reading the code

**Review by nfx — CHANGES_REQUESTED** [review]
Don't mark review comments as resolved unless you have really addressed them

File: src/databricks/labs/ucx/hive_metastore/hms_lineage.py (hunk 1 above)
Comment (nfx) [reply src/databricks/labs/ucx/hive_metastore/hms_lineage.py]: Keep scripts separate, yes. But don't add lineage enabler if it's already there. Manual reproduction: run installer twice and verify only one script created by ucx

Comment (dipankarkush-db) [reply src/databricks/labs/ucx/hive_metastore/hms_lineage.py]: Hi @nfx - This is what the code does -
if spark config present and the init script is enabled - skips creating a new one
If spark config present but the init script is disabled - asks if the user wants to enable it and enables/leave as it is based on user response
If spark config is not present then creates a new enabled global init script.

File: src/databricks/labs/ucx/runtime.py (hunk 2 above)
Comment (dipankarkush-db) [reply src/databricks/labs/ucx/runtime.py]: added in install.py

File: src/databricks/labs/ucx/hive_metastore/hms_lineage.py (hunk 3 above)
Comment (dipankarkush-db) [reply src/databricks/labs/ucx/hive_metastore/hms_lineage.py]: added.

File: src/databricks/labs/ucx/hive_metastore/hms_lineage.py (hunk 4 above)
Comment (dipankarkush-db) [reply src/databricks/labs/ucx/hive_metastore/hms_lineage.py]: removed.

File: src/databricks/labs/ucx/hive_metastore/hms_lineage.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+
+class HiveMetastoreLineageEnabler:
+    def __init__(self, ws: WorkspaceClient):
+        self._ws = ws
+
+    def check_lineage_spark_config_exists(self) -> GlobalInitScriptDetailsWithContent:
+        for script in self._ws.global_init_scripts.list():
+            gscript = self._ws.global_init_scripts.get(script_id=script.script_id)
+            if gscript:
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/hms_lineage.py:15]: Nit: Too much level of nesting, rewrite with "if gscript is None: continue"

**Review by nfx — APPROVED** [review]

File: src/databricks/labs/ucx/hive_metastore/hms_lineage.py (hunk 5 above)
Comment (dipankarkush-db) [reply src/databricks/labs/ucx/hive_metastore/hms_lineage.py]: Took care of the same.

File: src/databricks/labs/ucx/hive_metastore/hms_lineage.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+    def check_lineage_spark_config_exists(self) -> GlobalInitScriptDetailsWithContent:
+        for script in self._ws.global_init_scripts.list():
+            gscript = self._ws.global_init_scripts.get(script_id=script.script_id)
+            if not gscript:
+                continue
+            if "spark.databricks.dataLineage.enabled" in base64.b64decode(gscript.script).decode("utf-8"):
+                return gscript
+
+    def _get_init_script_content(self):
```
Comment (FastLee) [inline src/databricks/labs/ucx/hive_metastore/hms_lineage.py:20]: Keep the init script content in a file.

File: src/databricks/labs/ucx/install.py [hunk 7]
```diff
@@ -104,12 +105,51 @@ def run(self):
...
         self._create_dashboards()
         self._create_jobs()
         readme = f'{self._notebook_link(f"{self._install_folder}/README.py")}'
         msg = f"Installation completed successfully! Please refer to the {readme} notebook for next steps."
         logger.info(msg)
 
+    def _install_spark_config_for_hms_lineage(self):
+        if (
+            self._prompts
+            and self._question(
```
Comment (FastLee) [inline src/databricks/labs/ucx/install.py:118]: Check first if the script exists, before asking the question.

**Review by FastLee — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/hive_metastore/hms_lineage.py (hunk 6 above)
Comment (dipankarkush-db) [reply src/databricks/labs/ucx/hive_metastore/hms_lineage.py:24]: Hi @FastLee - took care of this.

File: src/databricks/labs/ucx/install.py (hunk 7 above)
Comment (dipankarkush-db) [reply src/databricks/labs/ucx/install.py]: Hi @FastLee - took care of this.

**Review by FastLee — APPROVED** [review]

**Pull request merged**